### PR TITLE
feat(core): Implement undo groups

### PR DESCRIPTION
**Pull request by elkowar**
This PR properly implements undo groups as well as undo and redo functionality

### CODE REVIEW

File: crates/bazed-core/src/buffer.rs [hunk 1]
```diff
@@ -165,24 +185,74 @@ impl Buffer {
...
+            let old_head_rev = self.engine.get_head_rev_id();
+
+            self.engine
+                .undo(self.undo_history.currently_undone().clone());
+            self.text = self.engine.get_head().clone();
+
+            match self.engine.try_delta_rev_head(old_head_rev.token()) {
+                Ok(delta) => self.marks.apply_delta(&delta),
+                Err(err) => {
+                    tracing::warn!("Error generating delta after redo: {err}");
```
Comment (buffet) [inline crates/bazed-core/src/buffer.rs:238]: `tracing::error`?

File: crates/bazed-core/src/buffer/undo_history.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+// edit[new_group=t] => id = 1, history = [0, 1         ], idx = 1, undone = {}
+// edit[new_group=f] => id = 1, history = [0, 1         ], idx = 1, undone = {}
+// edit[new_group=t] => id = 2, history = [0, 1, 2      ], idx = 2, undone = {}
+// edit[new_group=f] => id = 2, history = [0, 1, 2      ], idx = 2, undone = {}
+// edit[new_group=t] => id = 3, history = [0, 1, 2, 3   ], idx = 3, undone = {}
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 2, undone = { 3 }
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 1, undone = { 3, 2 }
+// undo              => id = 3, history = [0, 1, 2, 3   ], idx = 1, undone = { 3, 2 }
+// redo              => id = 3, history = [0, 1, 2, 3   ], idx = 2, undone = { 3 }
+// edit[new_group=t] => id = 4, history = [0, 1, 2,    4], idx = 3, undone = { 3 }
```
Comment (buffet) [inline crates/bazed-core/src/buffer/undo_history.rs:15]: These could be module docs or on the struct for easier discoverability.

**Review by buffet — CHANGES_REQUESTED** [review]

File: crates/bazed-core/src/buffer.rs [hunk 3]
```diff
@@ -153,7 +173,7 @@ impl Buffer {
             builder.replace(mark, text.clone());
         }
         let delta = builder.build();
-        self.commit_delta(delta);
+        self.commit_delta(delta, EditType::Insert);
```
[the remark this answers is not in the document]
Comment (elkowar) [reply crates/bazed-core/src/buffer.rs:176]: Very good point tho, that'll definitely need to exist, and in that case the api can actually be simplified a bit, lemme implement

File: crates/bazed-core/src/buffer/undo_history.rs (hunk 2 above)
Comment (elkowar) [reply crates/bazed-core/src/buffer/undo_history.rs]: Done!

File: crates/bazed-core/src/buffer.rs (hunk 1 above)
Comment (elkowar) [reply crates/bazed-core/src/buffer.rs]: also done

File: crates/bazed-core/src/buffer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (elkowar) [reply crates/bazed-core/src/buffer.rs:177]: Done!

File: crates/bazed-core/Cargo.toml [hunk 4]
```diff
@@ -22,3 +22,4 @@ tracing.workspace = true
 derive_more.workspace = true
 maplit.workspace = true
 tap.workspace = true
+tracing-subscriber.workspace = true
```
[the remark this answers is not in the document]
Comment (elkowar) [reply crates/bazed-core/Cargo.toml:27]: Done!

**Review by buffet — APPROVED** [review]
LGTM!

**Pull request merged**